tools/analysis_tools/confusion_matrix.py

```diff
@@ -0,0 +1,250 @@
+    parser.add_argument(
+        '--color-theme',
+        default='plasma',
+        help='theme of the matrix color map')
+    parser.add_argument(
+        '--score-thr',
+        type=float,
+        default=0,
+        help='score threshold to filter detection bboxes')
+    parser.add_argument(
+        '--tp-iou-thr',
+        type=float,
+        default=0.5,
+        help='IoU threshold to be considered as matched')
+    parser.add_argument(
+        '--nms-iou-thr',
+        type=float,
+        default=None,
+        help='nms IoU threshold, only applied when users want to change the'
+        'nms IoU threshold.')
+    parser.add_argument(
+        '--cfg-options',
+        nargs='+',
+        action=DictAction,
+        help='override some settings in the used config, the key-value pair '
+        'in xxx=yyy format will be merged into config file. If the value to '
+        'be overwritten is a list, it should be like key="[a,b]" or key=a,b '
+        'It also allows nested list/tuple values, e.g. key="[(a,b),(c,d)]" '
+        'Note that the quotation marks are necessary and that no white space '
+        'is allowed.')
+    args = parser.parse_args()
+    return args
+
+
+def calculate_confusion_matrix(dataset,
+                               results,
+                               score_thr=0,
+                               nms_iou_thr=None,
+                               tp_iou_thr=0.5):
+    """Calculate the confusion matrix.
+
+    Args:
+        dataset (Dataset): Test or val dataset.
+        results (list[ndarray]): A list of detection results in each image.
+        score_thr (float|optional): Score threshold to filter bboxes.
+            Default: 0.
+        nms_iou_thr (float|optional): nms IoU threshold, the detection results
+            have done nms in the detector, only applied when users want to
+            change the nms IoU threshold. Default: None.
+        tp_iou_thr (float|optional): IoU threshold to be considered as matched.
+            Default: 0.5.
+    """
+    num_classes = len(dataset.CLASSES)
+    confusion_matrix = np.zeros(shape=[num_classes + 1, num_classes + 1])
+    assert len(dataset) == len(results)
+    prog_bar = mmcv.ProgressBar(len(results))
+    for idx, per_img_res in enumerate(results):
+        if isinstance(per_img_res, tuple):
+            res_bboxes, _ = per_img_res
+        else:
+            res_bboxes = per_img_res
+        ann = dataset.get_ann_info(idx)
+        gt_bboxes = ann['bboxes']
+        labels = ann['labels']
+        analyze_per_img_dets(confusion_matrix, gt_bboxes, labels, res_bboxes,
+                             score_thr, tp_iou_thr, nms_iou_thr)
+        prog_bar.update()
+    per_label_sums = confusion_matrix.sum(axis=1)[:, np.newaxis]
+    normalized_matrix = \
+        confusion_matrix.astype('float') / per_label_sums * 100
+    return normalized_matrix
+
+
+def analyze_per_img_dets(confusion_matrix,
```

Comment:
Can we reuse the TP/FP calculation implemented in mean_ap?

Reply:
The function in mean_ap only returns the information of TP and FP, but in the confusion matrix, FN and category info are also needed. This implementation can do this in one loop.